src/shims/unix/linux/eventfd.rs

```diff
@@ -87,27 +148,41 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
     fn eventfd(&mut self, val: &OpTy<'tcx>, flags: &OpTy<'tcx>) -> InterpResult<'tcx, Scalar> {
         let this = self.eval_context_mut();
 
+        // eventfd is Linux specific.
+        this.assert_target_os("linux", "eventfd");
+
         let val = this.read_scalar(val)?.to_u32()?;
-        let flags = this.read_scalar(flags)?.to_i32()?;
+        let mut flags = this.read_scalar(flags)?.to_i32()?;
 
         let efd_cloexec = this.eval_libc_i32("EFD_CLOEXEC");
         let efd_nonblock = this.eval_libc_i32("EFD_NONBLOCK");
         let efd_semaphore = this.eval_libc_i32("EFD_SEMAPHORE");
 
-        if flags & (efd_cloexec | efd_nonblock | efd_semaphore) != flags {
-            throw_unsup_format!("eventfd: flag {flags:#x} is unsupported");
+        if flags & efd_semaphore == efd_semaphore {
+            throw_unsup_format!("eventfd: EFD_SEMAPHORE is unsupported");
         }
+
+        let mut is_nonblock = false;
+        // Unload the flag that we support.
+        // After unloading, flags != 0 means other flags are used.
         if flags & efd_cloexec == efd_cloexec {
-            // cloexec does nothing as we don't support `exec`
+            flags &= !efd_cloexec;
         }
         if flags & efd_nonblock == efd_nonblock {
-            // FIXME remember the nonblock flag
+            flags &= !efd_nonblock;
+            is_nonblock = true;
         }
-        if flags & efd_semaphore == efd_semaphore {
-            throw_unsup_format!("eventfd: EFD_SEMAPHORE is unsupported");
+        if flags != 0 {
+            let einval = this.eval_libc("EINVAL");
+            this.set_last_error(einval)?;
+            return Ok(Scalar::from_i32(-1));
```

Comment:
From my reading of the docs, this is how input errors are reported. Tho we can be more aggressive in marking this as invalid by stopping interpretation

Reply:
Most (all?) of our shims do throw_unsup when encountering a flag they do not understand. Returning an error code is uncommon. (https://github.com/rust-lang/miri/pull/3610 even switched one shim from an error code to throw_unsup.)

Or is it the case that we are checking all flags Linux supports, so any remaining flags are an invalid program, not a lack of Miri support?

Reply:
Yea, we exhausively check all flags, but new ones could be added in the future.

Let's use unsup then, people shouldn't be relying on invalid flags being "ok"

Reply:
>      [EINVAL]           Flags other than EFD_CLOEXEC, EFD_NONBLOCK, and EFD_SEMAPHORE are set in the flags argument.

[The man page](https://man.netbsd.org/eventfd.2) explicitly said that ``EINVAL`` will be thrown if there is unknown flags, (and this specification is uncommon). I think it is also ok to ``throw_unsup_format`` since we already do this in other shims. 

> Or is it the case that we are checking all flags Linux supports, so any remaining flags are an invalid program, not a lack of Miri support?

In this case, if there is other flags being added, it is indeed invalid program instead of lack of miri support.